### PR TITLE
feat(storage): Implement S3 Storage Adapter to be used with AWS S3 or compatible service like Fly's Tigris

### PR DESCRIPTION
**Pull request by dthyresson**
This PR adds a S3Storage adapter to Redwood Storage to be used with AWS S3 or compatible service like Fly's Tigris.

* `S3Storage` - the adapter that can read, save, remove and sign files to/from/in S3 storage
* `S3UrlSigner` - signed used by the Prisma extension to sign the file with the S3 sdk

Note: The `withSignedUrl` is now async just like the `withDataUri` is since for S3 the get signed url function in async. Therefore, some change are made to Prisma extension that if anyone is using, will need to be made async.

S3Storage also lets you define cache control headers.

When using S3Storage one does not need to use or have the `signedUrl` Redwood function to validate the url signation since this is all done by S3/Tigris. This save compute/requests on the app and ensures you get all the benefits of the S3 CDN to cache.

### CODE REVIEW

**Comment by Josh-Walker-GM** [comment]
I'm happy to review this but it'll be at least Monday (16th) before I do. We'll probably want this it it's own adapter package, like for mail handlers, since it brings in specific dependencies but I'm happy to do that sort of chore work since you've done the more important implementation part already here.

File: docs/docs/uploads.md
Comment (Tobbe) [inline docs/docs/uploads.md:483]: Does having `async` + `await` here make a difference? I thing you'll be returning an array of promises either way, but here you're creating an extra layer of promise wrapping (which the JIT compiler might be smart enough to remove, idk)

File: docs/docs/uploads.md
Comment (Tobbe) [inline docs/docs/uploads.md:735]:
```suggestion
- [S3Storage](https://github.com/redwoodjs/redwood/blob/main/packages/storage/src/adapters/S3Storage/S3Storage.ts) - this adapter stores files in AWS S3 or compatible service like Fly's Tigris. When using S3Storage you do not need a `signedUrl` endpoint to validate your signed urls because the S3 sdk can generate signed urls for you.
```

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (Tobbe) [inline packages/storage/src/__tests__/queryExtensions.test.ts:64]:
```suggestion
      // Mock the fs.mkdir function
      vi.mocked(fs.mkdir).mockResolvedValue(undefined)
```

Using `vi` instead of `jest`
Using `vi.mocked` instead of type casting

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (Tobbe) [inline packages/storage/src/__tests__/queryExtensions.test.ts:32]: `node:fs/promises` is already mocked further up the file. Isn't that enough?

File: packages/storage/src/adapters/S3Storage/S3Storage.ts
Comment (Tobbe) [inline packages/storage/src/adapters/S3Storage/S3Storage.ts:143]: I think it would be good to figure out the answer to this TODO before we merge this PR

File: packages/storage/src/adapters/S3Storage/S3UrlSigner.ts
Comment (Tobbe) [inline packages/storage/src/adapters/S3Storage/S3UrlSigner.ts:9]: We generally prefer to use JavaScript style `#private` if possible because that makes it a proper hard private class member.

**Review by Tobbe — CHANGES_REQUESTED** [review]

File: docs/docs/uploads.md
Comment (dthyresson) [reply docs/docs/uploads.md:483]: I don't know enough about promises to say one way or the other. I know that I need the `withSignedUrl` to be async so that the s3 sdk can call:

```
  async sign(fileLocation: string, expiresIn = 3600): Promise<string> {
    const command = new GetObjectCommand({
      Bucket: this.bucket,
      Key: fileLocation,
    })

    return await getSignedUrl(this.s3Client, command, { expiresIn })
  }
```

Without it didn't work.

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (dthyresson) [reply packages/storage/src/__tests__/queryExtensions.test.ts:64]: Ah. Thanks. That was AI gen'd and I changed to `vi` but I left that artifact.

File: packages/storage/src/adapters/S3Storage/S3UrlSigner.ts
Comment (dthyresson) [reply packages/storage/src/adapters/S3Storage/S3UrlSigner.ts:9]: I copied:

```ts
export class UrlSigner {
  private secret: string
  private endpoint: string

  constructor({ secret, endpoint }: SignedUrlSettings) {
    this.secret = secret
    this.endpoint = endpoint
```

Do both need to be changed @Tobbe ?

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (dthyresson) [reply packages/storage/src/__tests__/queryExtensions.test.ts:32]: If I remove I get ``` ❯ src/__tests__/queryExtensions.test.ts  (17 tests | 1 failed) 52ms
         × Query extensions > create > create will save files
           → Cannot read properties of undefined (reading 'mockResolvedValue')
       ✓ src/adapters/FileSystemStorage/FileSystemStorage.test.ts  (4 tests) 11ms
       ✓ src/adapters/MemoryStorage/MemoryStorage.test.ts  (6 tests) 3ms
      
      ⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯
      
       FAIL  src/__tests__/queryExtensions.test.ts > Query extensions > create > create will save files
      TypeError: Cannot read properties of undefined (reading 'mockResolvedValue')```

Let me see if can refactor.

Comment (dthyresson) [reply packages/storage/src/__tests__/queryExtensions.test.ts:32]: I wasn't able to but had comments added for purpose.

File: packages/storage/src/adapters/S3Storage/S3Storage.ts
Comment (dthyresson) [reply packages/storage/src/adapters/S3Storage/S3Storage.ts]: Checking status now

File: packages/storage/src/adapters/S3Storage/S3UrlSigner.ts
Comment (Tobbe) [reply packages/storage/src/adapters/S3Storage/S3UrlSigner.ts:9]: Nah, we can leave it for now and then fix it all later.

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (Tobbe) [inline packages/storage/src/__tests__/queryExtensions.test.ts:23]: @dthyresson This was all that was needed 🙂

File: packages/storage/src/adapters/FileSystemStorage/FileSystemStorage.ts
Comment (Tobbe) [inline packages/storage/src/adapters/FileSystemStorage/FileSystemStorage.ts:34]: This is using `fs` imported from `node:fs/promises`, so that's also what we should mock. I updated the test.

File: docs/docs/uploads.md
Comment (Tobbe) [reply docs/docs/uploads.md:483]: Yeah, I totally see why you had to make it async. No question there 🙂 Just don't think this need the extra async/await. But please don't just take my word for it. Try it out in a real project to make sure.

Comment (dthyresson) [reply docs/docs/uploads.md:483]: I did. Without the await in my describe pic and remove background job when I asked for the signedUrl it was undefined. When I made it async I got the url. Maybe within the service you do r need it but in a job I did so I could get the url and pass it along to fal as an image url reference.

Comment (dthyresson) [reply docs/docs/uploads.md:483]: Also the dataUri compute helper was always async so I just made both compute helpers work the same.

Comment (Tobbe) [reply docs/docs/uploads.md:483]: Yes, that all makes perfect sense. When you need to use the value for further computations/logic you need to await it first. But in the example you're just straight away returning it. That's why I don't think you need to await it. Just just passing the promise along.

This is the current example:
```ts
return filesViaRelation.map(async (file) => await file.withSignedUrl())
```

I'm just saying that in that particular case I think you could just have done this
```ts
return filesViaRelation.map((file) => file.withSignedUrl())
```
Because `withSignedUrl()` returns a promise and that promise is then used as the return value of the `map` callback. 

If you had to use `file` inside the callback after calling `withSignedUrl()` you would have had to await it first. Like this:
```ts
return filesViaRelation.map(async (file) => {
  const fileWithSignedUrl = await file.withSignedUrl()
  console.log('fileWithSignedUrl', fileWithSignedUrl)

  return fileWithSignedUrl
})
```

So yes, `withSignedUrl()` always has to be async. But whether you explicitly need to await it or not depends on how you use it.

I took an extra look inside `S3Storage.ts` and now see that I have similar feedback there. I think we should directly return the promise in `remove` and `sign` instead of first awaiting it and then returning.

But that requires some changes to `BaseStorageAdapter`. So maybe hold off on these changes until after we've discussed our current implementation on Tuesday

File: packages/storage/src/adapters/S3Storage/S3Storage.ts
Comment (Tobbe) [inline packages/storage/src/adapters/S3Storage/S3Storage.ts:46]: I don't think I've seen this before where you both extend and implement the same base class. What does that actually mean? What does it do?

Comment (dthyresson) [reply packages/storage/src/adapters/S3Storage/S3Storage.ts:46]: Not sure but again I followed the pattern established that the File and Memory adapters use in current main https://github.com/redwoodjs/redwood/blob/e384c39a9c201264d1b411149896e1afe6772fb0/packages/storage/src/adapters/FileSystemStorage/FileSystemStorage.ts#L12

Comment (dthyresson) [reply packages/storage/src/adapters/S3Storage/S3Storage.ts:46]: Now that I think about it the base adapter defines the common read etc but some also implement a remove and sign. So make implements makes sure the common methods are implemented and then extend says we have some new custom methods like sign or remove but still consider the class to be like the base adapter? Thus the extend?

File: packages/storage/src/__tests__/queryExtensions.test.ts
Comment (dthyresson) [reply packages/storage/src/__tests__/queryExtensions.test.ts:23]: ah. thanks. tobbAI > Cursor AI :)

**Comment by dthyresson** [comment]
I decided to mark this as a draft with some upcoming Storage refactoring. It makes more sense to add a S3 adapter after that happens.

**Comment by dthyresson** [comment]
Closing as we'll have a new interface for adapters soon.

**Pull request closed without merging**
